docs/review_02.03.22/review_02.03.22.md

```diff
@@ -0,0 +1,546 @@
+still in sync with `monerod`. Finally, if there is a sync error that is not a
+timeout, throws; if not timeout, warns in a log. Continues looping until
+there are users to scan transactions for.
+
+###### check_loop()
+
+Relevant comments:
+
+> Launches `thread_count` threads to run `scan_loop`, and then polls for active account changes in background
+
+```
+The algorithm here is extremely basic. Users are divided evenly amongst
+the configurable thread count, and grouped by scan height. If an old
+account appears, some accounts (grouped on that thread) will be delayed
+in processing waiting for that account to catch up. Its not the greatest,
+but this "will have to do" for the first cut.
+Its not expected that many people will be running
+"enterprise level" of nodes where accounts are constantly added.
+
+Another "issue" is that each thread works independently instead of more
+cooperatively for scanning. This requires a bit more synchronization, so
+was left for later. Its likely worth doing to reduce the number of
+transfers from the daemon, and the bottleneck on the writes into LMDB.
+
+If the active user list changes, all threads are stopped/joined, and
+everything is re-started.
+```
+
+Exactly as described, the first thing it does is sorts all users by 
+`scan_height`, so that it can send batches of users at a time into threads
+that execute `scan_loop()` over those users (so that users with similar heights
+are processed together). Then it enters a while loop that updates exchange rates
+over the exchange rate interval, and then perpetually checks for changes to
+the users stored in the `monero-lws` db. If it finds that there is a change
+in active users stored, `check_loop()` returns so that `lws::scanner::run()`
+can do another loop, collecting all active users again and re-entering
+`check_loop()` with the updated list.
+
+###### scan_loop()
+
+TO-DO
+
+#### /src/admin_main.cpp
+
+#### /src/CMakeLists.txt
+
+#### /src/config.cpp
+
+#### /src/config.h
+
+#### /src/error.cpp
+
+#### /src/error.h
+
+#### /src/fwd.h
+
+#### /src/options.h
+
+#### /src/wire.h
+
+#### /src/db
+
+##### /src/db/account.cpp
+
+##### /src/db/account.h
+
+##### /src/db/CMakeLists.txt
+
+##### /src/db/data.cpp
+
+##### /src/db/data.h
+
+##### /src/db/fwd.h
+
+##### /src/db/storage.cpp
+
+##### /src/db/storage.h
+
+##### /src/db/string.cpp
+
+##### /src/db/string.h
+
+#### /src/rpc
+
+##### /src/rpc/client.cpp
+
+##### /src/rpc/client.h
+
+##### /src/rpc/CMakeLists.txt
+
+##### /src/rpc/daemon_pub.cpp
+
+##### /src/rpc/daemon_pub.h
+
+##### /src/rpc/daemon_zmq.cpp
+
+##### /src/rpc/daemon_zmq.h
+
+##### /src/rpc/fwd.h
+
+##### /src/rpc/json.h
+
+##### /src/rpc/ligt_wallet.cpp
+
+##### /src/rpc/ligt_wallet.h
+
+##### /src/rpc/rates.cpp
+
+##### /src/rpc/rates.h
+
+#### /src/util
+
+#### /src/wire
+
+### /CMakeLists.txt
+
+## Suggestions
+
+- Consider indexing addresses by a hash of the private view key instead of just
```

Comment:
This is a decent suggestion. I _think_ the check for view_key -> view_public_key is always done before any DB lookups to prevent an obvious leak (that was my intent, but probably wasn't documented in the comments). Since LMDB is doing a sorted lookup, there might some issues with approved lws users recovering some lower order bits. The timing should depend on the number of accounts, the state of the file cache, state of the cpu cache, etc., but might be possible in some situations.

For those unable to follow - the recovery process should involve trying to determine where _your_ account is listed in the LMDB B-tree, which gives indirect information on the number of accounts and possible view-public key values of those accounts. Hashing would only "blind" the second case (the public-key values), and only when the attacker is attempting to recover values from nothing (i.e. if the attacker has some suspected account addresses, then the attacker can still hash+recover).

This is very pessimistic and must be weighed against having the first DB upgrade code in `monero-lws`. And if deemed important enough, its worth digging deeper to make other DB changes simultaneously to reduce amount of upgrade code.

Reply:
> Since LMDB is doing a sorted lookup, there might some issues with approved lws users recovering some lower order bits.

I was debating clarifying that this is what I meant. Just some amount of the earlier bits could be revealed in theory.

> This is very pessimistic and must be weighed against having the first DB upgrade code in monero-lws

Agreed, very well could be not worth implementing.

Reply:
I intend to switch to a constant time comparison (in the specified LMDB sort) for the account address. This doesn't require any changes to the stored data, and helps hide the public view key during searching.

I am also leaning towards `cryptographic_hash(salt || public_view_key)` (in addition to the above), but I haven't thought through all of the scenarios yet.